src/main/telemetry/ghst.c

```diff
@@ -0,0 +1,226 @@
+    sbufWriteU8(dst, 0x23);						// GHST_DL_PACK_STAT
+
+    if (telemetryConfig()->report_cell_voltage) {
+    	sbufWriteU16(dst, getBatteryAverageCellVoltage());		// units of 10mV
+    } else {
+    	sbufWriteU16(dst, getBatteryVoltage());
+    }
+    sbufWriteU16(dst, getAmperage());							// units of 10mA
+
+    sbufWriteU16(dst, getMAhDrawn());							// units of 10mAh
+
+    sbufWriteU8(dst, 0x00);						// Rx Voltage, units of 100mV (not passed from BF, added in Ghost Rx)
+
+    sbufWriteU8(dst, 0x00);			            // tbd1
+    sbufWriteU8(dst, 0x00);						// tbd2
+    sbufWriteU8(dst, 0x00);						// tbd3
+}
+
+#define BV(x)  (1 << (x))                       // bit value
+
+// schedule array to decide how often each type of frame is sent
+typedef enum {
+    GHST_FRAME_START_INDEX = 0,
+    GHST_FRAME_PACK_INDEX = GHST_FRAME_START_INDEX,			// Battery (Pack) data
+    GHST_SCHEDULE_COUNT_MAX
+} ghstFrameTypeIndex_e;
+
+static uint8_t ghstScheduleCount;
+static uint8_t ghstSchedule[GHST_SCHEDULE_COUNT_MAX];
+
+static void processGhst(void)
+{
+    static uint8_t ghstScheduleIndex = 0;
+
+    const uint8_t currentSchedule = ghstSchedule[ghstScheduleIndex];
+
+    sbuf_t ghstPayloadBuf;
+    sbuf_t *dst = &ghstPayloadBuf;
+
+    if (currentSchedule & BV(GHST_FRAME_PACK_INDEX)) {
+        ghstInitializeFrame(dst);
+        ghstFramePackTelemetry(dst);
+        ghstFinalize(dst);
+    }
+    ghstScheduleIndex = (ghstScheduleIndex + 1) % ghstScheduleCount;
+}
+
+void ghstScheduleDeviceInfoResponse(void)
+{
+    deviceInfoReplyPending = true;
+}
+
+void initGhstTelemetry(void)
+{
+    // If the GHST Rx driver is active, since tx and rx share the same pin, assume telemetry is enabled. 
+    ghstTelemetryEnabled = ghstRxIsActive();
+
+    if (!ghstTelemetryEnabled) {
+        return;
+    }
+
+    deviceInfoReplyPending = false;
+
+    int index = 0;
+    if ((isBatteryVoltageConfigured() && telemetryIsSensorEnabled(SENSOR_VOLTAGE))
+        || (isAmperageConfigured() && telemetryIsSensorEnabled(SENSOR_CURRENT | SENSOR_FUEL))) {
+        ghstSchedule[index++] = BV(GHST_FRAME_PACK_INDEX);
+    }
+    ghstScheduleCount = (uint8_t)index;
+ }
+
+bool checkGhstTelemetryState(void)
+{
+    return ghstTelemetryEnabled;
+}
+
+// Called periodically by the scheduler
+ void handleGhstTelemetry(timeUs_t currentTimeUs)
+{
+    static uint32_t ghstLastCycleTime;
+
+    if (!ghstTelemetryEnabled) {
+        return;
+    }
+
+    // Give the receiver a chance to send any outstanding telemetry data.
+    // This needs to be done at high frequency, to enable the RX to send the telemetry frame
+    // in between the RX frames.
+    ghstRxSendTelemetryData();
+
+    // Ready to send telemetry?
+    if (currentTimeUs >= ghstLastCycleTime + (GHST_CYCLETIME_US / ghstScheduleCount)) {
+        ghstLastCycleTime = currentTimeUs;
+        processGhst();
+    }
+
+}
+
+// only used for unit testing
```

Comment:
This should not be in here then. Please move into the unit tests.

Reply:
Removed for now